sklearn/tree/_criterion.pyx

```diff
@@ -283,7 +283,7 @@ cdef class ClassificationCriterion(Criterion):
 
     cdef void init(self, DOUBLE_t* y, SIZE_t y_stride,
                    DOUBLE_t* sample_weight, double weighted_n_samples,
-                   SIZE_t* samples, SIZE_t start, SIZE_t end) nogil:
+                   SIZE_t* samples, SIZE_t start, SIZE_t end) nogil except *:
```

Comment:
It would have been cleaner to use an int return code to make that internal memory allocation more explicit though.

@larsmans @jnothman do you have any opinion? Shall we change the API to add the return code and make the allocation explicit?